File: 6/merge_two_binary_trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def mergeTreesRecursive(self, t1: TreeNode, t2: TreeNode) -> TreeNode:
        if not t1:
            return t2
        if not t2:
            return t1
        
        t1.val += t2.val
        t1.left = self.mergeTrees(t1.left, t2.left)
        t1.right = self.mergeTrees(t1.right, t2.right)
        return t1
    
    def mergeTrees(self, t1: TreeNode, t2: TreeNode) -> TreeNode:
        if not t1:
            return t2
        if not t2:
            return t1
        
        stack = [(t1, t2)]
        while stack:
            n1, n2 = stack.pop()
            if not n1 or not n2:
                continue
            n1.val += n2.val
            
            if not n1.left:
                n1.left = n2.left
            else:
                stack.append((n1.left, n2.left))
            if not n1.right:
                n1.right = n2.right
            else:
                stack.append((n1.right, n2.right))
        return t1

File: 6/merge_two_binary_trees.go
package main

import "fmt"

// TreeNode is the struct for tree node
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func mergeTrees(t1 *TreeNode, t2 *TreeNode) *TreeNode {
	if *t1 == (TreeNode{}) {
		return t2
	}
	if *t2 == (TreeNode{}) {
		return t1
	}
	stack := [][]*TreeNode{[]*TreeNode{t1, t2}}
	for len(stack) > 0 {
		n1 := stack[len(stack)-1][0]
		n2 := stack[len(stack)-1][1]
		stack = stack[0 : len(stack)-1]
		if *n1 == (TreeNode{}) || *n2 == (TreeNode{}) {
			continue
		}
		n1.Val += n2.Val
		if *(n1.Left) == (TreeNode{}) {
			n1.Left = n2.Left
		} else {
			stack = append(stack, []*TreeNode{n1.Left, n2.Left})
		}
		if *(n1.Right) == (TreeNode{}) {
			n1.Right = n2.Right
		} else {
			stack = append(stack, []*TreeNode{n1.Right, n2.Right})
		}
	}
	return t1
}

func printTree(root *TreeNode) {
	q := []*TreeNode{root}
	res := make([]int, 0)
	for len(q) > 0 {
		newQ := make([]*TreeNode, 0)
		for _, node := range q {
			res = append(res, node.Val)
			if *(node.Left) != (TreeNode{}) {
				newQ = append(newQ, node.Left)
			}
			if *(node.Right) != (TreeNode{}) {
				newQ = append(newQ, node.Right)
			}
		}
		q = newQ
	}
	fmt.Println(res)
}

func main() {
	t1 := TreeNode{1, &TreeNode{3, &TreeNode{5, &TreeNode{}, &TreeNode{}}, &TreeNode{}}, &TreeNode{2, &TreeNode{}, &TreeNode{}}}
	t2 := TreeNode{2, &TreeNode{1, &TreeNode{}, &TreeNode{4, &TreeNode{}, &TreeNode{}}}, &TreeNode{3, &TreeNode{}, &TreeNode{7, &TreeNode{}, &TreeNode{}}}}
	printTree(&t1)
	printTree(&t2)
	printTree(mergeTrees(&t1, &t2))
}
